Support scopes on @Bean methods in spring-di

From spring Javadoc (https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/annotation/Bean.java#L82):

> `@Scope`  [on method annotated with  `@Bean` ] changes the bean's scope from singleton to the specified scope" 

But the current implementation of SpringDiProcessor doesn't conform to this:

1. `@Dependent` scope is added to all beans produced by `@Bean` methods in `@Configuration` classes, so following #2178 in Quarkus 0.14.0+ it generates an error:

```
@Configuration
public class MyConfig {

	@Bean
	@Singleton // Causes "multiple scope type annotations" error
	public SomeBean myBean() {
		return new SomeBean();
	}

	private static class SomeBean {
	}
}
```

Actual behavior:
```
Caused by: javax.enterprise.inject.spi.DefinitionException: Producer method org.acme.quickstart.MyConfig$SomeBean myBean() declares multiple scope type annotations: javax.inject.Singleton, javax.enterprise.context.Dependent
	at io.quarkus.arc.processor.Beans.multipleScopesFound(Beans.java:281)
	at io.quarkus.arc.processor.Beans.createProducerMethod(Beans.java:175)
	at io.quarkus.arc.processor.BeanDeployment.findBeans(BeanDeployment.java:696)
	at io.quarkus.arc.processor.BeanDeployment.<init>(BeanDeployment.java:142)
	at io.quarkus.arc.processor.BeanProcessor.process(BeanProcessor.java:153)
	at io.quarkus.arc.deployment.ArcAnnotationProcessor.build(ArcAnnotationProcessor.java:282)
```
Expected behavior: the default scope should be singleton (not dependent), and it should also be allowed to make it explicit with `@javax.inject.Singleton` without causing "multiple scope type annotations" error.

2. Using `@org.springframework.context.annotation.Scope` causes a NPE: 

```
@Bean
@Scope("singleton") // NPE
public SomeBean myBean() {
	return new SomeBean();
}
```

Actual behavior:
```
Caused by: java.lang.NullPointerException
	at io.quarkus.spring.di.deployment.SpringDIProcessor.getScope(SpringDIProcessor.java:190)
	at io.quarkus.spring.di.deployment.SpringDIProcessor.getAnnotationsToAdd(SpringDIProcessor.java:297)
	at io.quarkus.spring.di.deployment.SpringDIProcessor.lambda$beanTransformer$1(SpringDIProcessor.java:126)
	at io.quarkus.arc.processor.AnnotationStore.transform(AnnotationStore.java:116)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1688)
	at io.quarkus.arc.processor.AnnotationStore.getAnnotations(AnnotationStore.java:70)
	at io.quarkus.arc.processor.AnnotationStore.hasAnnotation(AnnotationStore.java:94)
	at io.quarkus.arc.processor.BeanDeployment.findBeans(BeanDeployment.java:599)
	at io.quarkus.arc.processor.BeanDeployment.<init>(BeanDeployment.java:142)
	at io.quarkus.arc.processor.BeanProcessor.process(BeanProcessor.java:153)
	at io.quarkus.arc.deployment.ArcAnnotationProcessor.build(ArcAnnotationProcessor.java:282)
```
Expected behavior: scope should be singleton.

3.  Other Spring scopes (`org.springframework.web.context.annotation.RequestScope`) do not seem taken into account ?

```
@Bean
@RequestScope
public SomeBean myBean() {
	return new SomeBean();
}
```

Actual behavior:
```
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.quickstart.MyBean and qualifiers [@Default]
	- java member: org.acme.quickstart.GreetingService#<init>()
	- declared on CLASS bean [types=[org.acme.quickstart.GreetingService, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.quickstart.GreetingService]
	at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:378)
	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:366)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:286)
```
Expected behavior: the bean should be seen, with scope javax.enterprise.context.RequestScoped

**To Reproduce**
Steps to reproduce the behavior:
1. Add configuration class into a module with empty META-INF/beans.xml
2. Use spring-di quarkus extension
3. Run a @QuarkusTest or Maven build

**Environment:**
 - Output of `uname -a` or `ver`: Microsoft Windows [version 10.0.17134.765]
 - Output of `java -version`: jdk1.8.0_181 Java HotSpot(TM) 64-Bit Server VM
 - GraalVM version (if different from Java): N/A
 - Quarkus version or git rev: 0.16.1

**Additional context**
I was looking at contributing a quick fix to SpringDiProcessor, but it appears more complex I initially thought:
- https://github.com/quarkusio/quarkus/blob/0.16.1/extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java#L187 : Jandex ClassInfo#annotations() returns all annotations (not only class-level) so this line seems wrong, but changing the behavior to instead loop on ClassInfo#classAnnotations() breaks some existing unit tests.
- in Jandex annotations are not stored in parent class AnnotationTarget (that's sad !), so the existing scope/stereotype determination logic will need be duplicated/refactored to support ClassInfo or MethodInfo.

--- a/extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java
+++ b/extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java
@@ -162,18 +162,24 @@ public class SpringDIProcessor {
     }
 
     /**
-     * Translate spring built in scope identifiers to CDI scopes.
+     * Translate spring built-in scope identifiers to CDI scopes.
      *
-     * @param clazz The class declaring the @Scope
+     * @param target The annotated element declaring the @Scope
      * @return A CDI built in (or session) scope that mostly matches
      *         the spring one. Websocket scope is currently mapped to @Dependant
      *         and spring custom scopes are not currently handled.
      */
-    private DotName getScope(final ClassInfo clazz) {
-        if (!clazz.annotations().containsKey(SPRING_SCOPE_ANNOTATION)) {
-            return null;
+    private DotName getScope(final AnnotationTarget target) {
+        AnnotationValue value = null;
+        if (target.kind() == AnnotationTarget.Kind.CLASS) {
+            if (target.asClass().classAnnotation(SPRING_SCOPE_ANNOTATION) != null) {
+                value = target.asClass().classAnnotation(SPRING_SCOPE_ANNOTATION).value();
+            }
+        } else if (target.kind() == AnnotationTarget.Kind.METHOD) {
+            if (target.asMethod().hasAnnotation(SPRING_SCOPE_ANNOTATION)) {
+                value = target.asMethod().annotation(SPRING_SCOPE_ANNOTATION).value();
+            }
         }
-        final AnnotationValue value = clazz.classAnnotation(SPRING_SCOPE_ANNOTATION).value();
         if (value != null) {
             switch (value.asString()) {
                 case "singleton":
@@ -372,10 +378,17 @@ public class SpringDIProcessor {
                         CDI_PRODUCES_ANNOTATION,
                         target,
                         Collections.emptyList()));
-                annotationsToAdd.add(create(
-                        CDI_DEPENDENT_ANNOTATION,
-                        target,
-                        Collections.emptyList()));
+                DotName declaredScope = getScope(methodInfo);
+                // Non-spring specific annotations are processed by Arc
+                if (declaredScope == null && !methodInfo.hasAnnotation(CDI_SINGLETON_ANNOTATION)) {
+                    declaredScope = CDI_SINGLETON_ANNOTATION; // implicit default scope in spring
+                }
+                if (declaredScope != null) {
+                    annotationsToAdd.add(create(
+                            declaredScope,
+                            target,
+                            Collections.emptyList()));
+                }
 
                 //check if the spring annotation defines a name for the bean
                 final AnnotationValue beanNameAnnotationValue = methodInfo.annotation(BEAN_ANNOTATION).value("name");

--- a/extensions/spring-di/deployment/src/test/java/io/quarkus/spring/di/deployment/SpringDIProcessorTest.java
+++ b/extensions/spring-di/deployment/src/test/java/io/quarkus/spring/di/deployment/SpringDIProcessorTest.java
@@ -27,13 +27,17 @@ import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
 import org.jboss.jandex.Indexer;
+import org.jboss.jandex.MethodInfo;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Scope;
 import org.springframework.stereotype.Component;
 import org.springframework.stereotype.Service;
 
 import io.quarkus.arc.processor.BeanArchives;
+import io.quarkus.arc.processor.DotNames;
 import io.quarkus.deployment.util.IoUtil;
 
 /**
@@ -42,9 +46,10 @@ import io.quarkus.deployment.util.IoUtil;
 class SpringDIProcessorTest {
 
     final SpringDIProcessor processor = new SpringDIProcessor();
-    final IndexView index = getIndex(PrototypeService.class, RequestService.class,
-            UndeclaredService.class, ConflictService.class, ConflictBean.class, RequestBean.class, UndeclaredBean.class,
-            ConflictStereotypeBean.class, NamedBean.class, OverrideConflictStereotypeBean.class, ConflictNamedBean.class);
+    final IndexView index = getIndex(PrototypeService.class, RequestService.class, UndeclaredService.class,
+            ConflictService.class, ConflictBean.class, RequestBean.class, UndeclaredBean.class,
+            ConflictStereotypeBean.class, NamedBean.class, OverrideConflictStereotypeBean.class,
+            ConflictNamedBean.class);
 
     @Test
     public void springStereotypeScopes() {
@@ -58,10 +63,9 @@ class SpringDIProcessorTest {
                 Collections.singleton(DotName.createSimple(Dependent.class.getName())));
         expected.put(DotName.createSimple(RequestService.class.getName()),
                 Collections.singleton(DotName.createSimple(RequestScoped.class.getName())));
-        expected.put(DotName.createSimple(UndeclaredService.class.getName()),
-                Collections.emptySet());
-        expected.put(DotName.createSimple(ConflictService.class.getName()),
-                setOf(DotName.createSimple(Dependent.class.getName()), DotName.createSimple(RequestScoped.class.getName())));
+        expected.put(DotName.createSimple(UndeclaredService.class.getName()), Collections.emptySet());
+        expected.put(DotName.createSimple(ConflictService.class.getName()), setOf(
+                DotName.createSimple(Dependent.class.getName()), DotName.createSimple(RequestScoped.class.getName())));
         assertEquals(expected, scopes);
     }
 
@@ -146,6 +150,53 @@ class SpringDIProcessorTest {
         });
     }
 
+    @Test
+    public void getAnnotationsToAddBeanMethodDefaultsToSingleton() {
+        final Map<DotName, Set<DotName>> scopes = processor.getStereotypeScopes(index);
+        final MethodInfo target = index.getClassByName(DotName.createSimple(BeanConfig.class.getName()))
+                .method("singletonBean");
+
+        final Set<AnnotationInstance> ret = processor.getAnnotationsToAdd(target, scopes, null);
+
+        final Set<AnnotationInstance> expected = setOf(
+                AnnotationInstance.create(DotName.createSimple(Singleton.class.getName()), target,
+                        Collections.emptyList()),
+                AnnotationInstance.create(DotNames.PRODUCES, target, Collections.emptyList()));
+        assertEquals(expected, ret);
+    }
+
+    /**
+     * Do not add javax.inject.Singleton, as it is not spring-specific and Arc processor already handles it.
+     * Otherwise it would cause "declares multiple scope type annotations" error.
+     */
+    @Test
+    public void getAnnotationsToAddBeanMethodExplicitSingleton() {
+        final Map<DotName, Set<DotName>> scopes = processor.getStereotypeScopes(index);
+        final MethodInfo target = index.getClassByName(DotName.createSimple(BeanConfig.class.getName()))
+                .method("explicitSingletonBean");
+
+        final Set<AnnotationInstance> ret = processor.getAnnotationsToAdd(target, scopes, null);
+
+        final Set<AnnotationInstance> expected = setOf(
+                AnnotationInstance.create(DotNames.PRODUCES, target, Collections.emptyList()));
+        assertEquals(expected, ret);
+    }
+
+    @Test
+    public void getAnnotationsToAddBeanMethodWithScope() {
+        final Map<DotName, Set<DotName>> scopes = processor.getStereotypeScopes(index);
+        final MethodInfo target = index.getClassByName(DotName.createSimple(BeanConfig.class.getName()))
+                .method("requestBean");
+
+        final Set<AnnotationInstance> ret = processor.getAnnotationsToAdd(target, scopes, null);
+
+        final Set<AnnotationInstance> expected = setOf(
+                AnnotationInstance.create(DotName.createSimple(RequestScoped.class.getName()), target,
+                        Collections.emptyList()),
+                AnnotationInstance.create(DotNames.PRODUCES, target, Collections.emptyList()));
+        assertEquals(expected, ret);
+    }
+
     private IndexView getIndex(final Class<?>... classes) {
         final Indexer indexer = new Indexer();
         for (final Class<?> clazz : classes) {
@@ -222,5 +273,26 @@ class SpringDIProcessorTest {
     @ConflictService
     @Scope("application")
     public class OverrideConflictStereotypeBean {
+    }
+
+    @Configuration
+    public class BeanConfig {
+
+        @Bean
+        public UndeclaredBean singletonBean() {
+            return new UndeclaredBean();
+        }
+
+        @Bean
+        @Scope(value = "request")
+        public UndeclaredBean requestBean() {
+            return new UndeclaredBean();
+        }
+
+        @Bean
+        @Singleton
+        public UndeclaredBean explicitSingletonBean() {
+            return new UndeclaredBean();
+        }
     }
 }

--- a/integration-tests/spring-di/pom.xml
+++ b/integration-tests/spring-di/pom.xml
@@ -38,6 +38,11 @@
             <artifactId>spring-context</artifactId>
             <version>${spring.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-web</artifactId>
+            <version>${spring.version}</version>
+        </dependency>
 
         <dependency>
             <groupId>io.quarkus</groupId>

--- a/integration-tests/spring-di/src/main/java/io/quarkus/it/spring/AppConfiguration.java
+++ b/integration-tests/spring-di/src/main/java/io/quarkus/it/spring/AppConfiguration.java
@@ -1,8 +1,11 @@
 package io.quarkus.it.spring;
 
+import javax.inject.Singleton;
+
 import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.web.context.annotation.RequestScope;
 
 @Configuration
 public class AppConfiguration {
@@ -11,5 +14,30 @@ public class AppConfiguration {
     public StringFunction capitalizer(@Qualifier("dumb") Dummy notUsedJustMakingSureDIInMethodsWorks,
             OtherDummy alsoNotUsed) {
         return String::toUpperCase;
+    }
+
+    @Bean
+    @Singleton
+    public SingletonBean explicitSingletonBean() {
+        return new SingletonBean();
+    }
+
+    @Bean
+    public SingletonBean implicitSingletonBean() {
+        return new SingletonBean();
+    }
+
+    @Bean
+    @RequestScope
+    public AnotherRequestBean requestBean() {
+        return new AnotherRequestBean();
+    }
+
+    private static class SingletonBean {
+
+    }
+
+    private static class AnotherRequestBean {
+
     }
 }

